File: ALittle/Module/ALittleIDE/Other/GameLibrary/SEngine/Script/Utility/HttpSystem.lua
-- ALittle Generate Lua
module("ALittle", package.seeall)

local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


assert(ALittle.IHttpReceiverNative, " extends class:ALittle.IHttpReceiverNative is nil")
HttpReceiverNative = Lua.Class(ALittle.IHttpReceiverNative, "ALittle.HttpReceiverNative")

function HttpReceiverNative:Close(http_id)
	__CPPAPI_ServerSchedule:HttpClose(http_id)
end

function HttpReceiverNative:SendString(http_id, content)
	__CPPAPI_ServerSchedule:HttpSendString(http_id, content)
end

function HttpReceiverNative:SendFile(http_id, file_path, start_size)
	__CPPAPI_ServerSchedule:HttpSendFile(http_id, file_path, start_size)
end

local HttpReceiver = Lua.Template(HttpReceiverTemplate, "ALittle.HttpReceiverTemplate<ALittle.HttpReceiverNative>", HttpReceiverNative);
HttpSystem = Lua.Class(nil, "ALittle.HttpSystem")

function HttpSystem:Ctor()
	___rawset(self, "_request_map", {})
	___rawset(self, "_id_creator", SafeIDCreator())
end

function HttpSystem.HandleHttpTask(http_id, callback, value_map)
	local client = HttpReceiver(http_id)
	local error, result = Lua.TCall(callback, client, value_map)
	if error ~= nil then
		local map = {}
		map["error"] = error
		client:SendString(json.encode(map))
		return
	end
	if result == nil then
		local map = {}
		map["error"] = "接口没有返回任何内容"
		client:SendString(json.encode(map))
		return
	end
	client:SendString(json.encode(result))
end
HttpSystem.HandleHttpTask = Lua.CoWrap(HttpSystem.HandleHttpTask)

function HttpSystem.HandleHttpDownloadTask(http_id, callback, value_map)
	local client = HttpReceiver(http_id)
	local error, file_path, start_size = Lua.TCall(callback, client, value_map)
	if error ~= nil then
		client:Close()
		return
	end
	if file_path == nil then
		client:Close()
		return
	end
	client:SendFile(file_path, start_size)
end
HttpSystem.HandleHttpDownloadTask = Lua.CoWrap(HttpSystem.HandleHttpDownloadTask)

function HttpSystem:Get(url)
local ___COROUTINE = coroutine.running()
	if ___COROUTINE == nil then
		return "当前不是协程", nil
	end
	local id = self._id_creator:CreateID()
	self._request_map[id] = ___COROUTINE
	__CPPAPI_ServerSchedule:HttpGet(id, url)
	return coroutine.yield()
end

function HttpSystem:PostJson(url, content)
local ___COROUTINE = coroutine.running()
	if ___COROUTINE == nil then
		return "当前不是协程", nil
	end
	local id = self._id_creator:CreateID()
	self._request_map[id] = ___COROUTINE
	__CPPAPI_ServerSchedule:HttpPost(id, url, "application/json", json.encode(content))
	return coroutine.yield()
end

function HttpSystem:HandleHttpSucceed(id, response)
	self._id_creator:ReleaseID(id)
	local thread = self._request_map[id]
	self._request_map[id] = nil
	local result, reason = coroutine.resume(thread, nil, response)
	if result ~= true then
		Error(reason)
	end
end

function HttpSystem:HandleHttpFailed(id, reason)
	self._id_creator:ReleaseID(id)
	local thread = self._request_map[id]
	self._request_map[id] = nil
	local result, reason = coroutine.resume(thread, reason, nil)
	if result ~= true then
		Error(reason)
	end
end

_G.A_HttpSystem = HttpSystem()

File: ALittle/Module/ALittleIDE/Other/GameLibrary/SEngine/JSScript/Utility/HttpSystem.js
{
if (typeof ALittle === "undefined") ALittle = {};


if (ALittle.IHttpReceiverNative === undefined) throw new Error(" extends class:ALittle.IHttpReceiverNative is undefined");
ALittle.HttpReceiverNative = JavaScript.Class(ALittle.IHttpReceiverNative, {
	Close : function(http_id) {
		__CPPAPI_ServerSchedule.HttpClose(http_id);
	},
	SendString : function(http_id, content) {
		__CPPAPI_ServerSchedule.HttpSendString(http_id, content);
	},
	SendFile : function(http_id, file_path, start_size) {
		__CPPAPI_ServerSchedule.HttpSendFile(http_id, file_path, start_size);
	},
}, "ALittle.HttpReceiverNative");

let HttpReceiver = JavaScript.Template(ALittle.HttpReceiverTemplate, "ALittle.HttpReceiverTemplate<ALittle.HttpReceiverNative>", ALittle.HttpReceiverNative);
ALittle.HttpSystem = JavaScript.Class(undefined, {
	Ctor : function() {
		this._request_map = new Map();
		this._id_creator = ALittle.NewObject(ALittle.SafeIDCreator);
	},
	HandleHttpTask : async function(http_id, callback, value_map) {
		let client = ALittle.NewObject(ALittle.HttpReceiver, http_id);
		let [error, result] = await (async function() { try { let ___VALUE = await callback.call(client, value_map); return [undefined, ___VALUE]; } catch (___ERROR) { return [___ERROR.message]; } })();
		if (error !== undefined) {
			let map = {};
			map["error"] = error;
			client.SendString(lua.json.encode(map));
			return;
		}
		if (result === undefined) {
			let map = {};
			map["error"] = "接口没有返回任何内容";
			client.SendString(lua.json.encode(map));
			return;
		}
		client.SendString(lua.json.encode(result));
	},
	HandleHttpDownloadTask : async function(http_id, callback, value_map) {
		let client = ALittle.NewObject(ALittle.HttpReceiver, http_id);
		let [error, file_path, start_size] = await (async function() { try { let ___VALUE = await callback.call(client, value_map); ___VALUE.splice(0, 0, undefined);  return ___VALUE; } catch (___ERROR) { return [___ERROR.message]; } })();
		if (error !== undefined) {
			client.Close();
			return;
		}
		if (file_path === undefined) {
			client.Close();
			return;
		}
		client.SendFile(file_path, start_size);
	},
	Get : function(url) {
		return new Promise((function(___COROUTINE, ___) {
			if (___COROUTINE === undefined) {
				___COROUTINE(["当前不是协程", undefined]); return;
			}
			let id = this._id_creator.CreateID();
			this._request_map.set(id, ___COROUTINE);
			__CPPAPI_ServerSchedule.HttpGet(id, url);
			return;
			___COROUTINE();
		}).bind(this));
	},
	PostJson : function(url, content) {
		return new Promise((function(___COROUTINE, ___) {
			if (___COROUTINE === undefined) {
				___COROUTINE(["当前不是协程", undefined]); return;
			}
			let id = this._id_creator.CreateID();
			this._request_map.set(id, ___COROUTINE);
			__CPPAPI_ServerSchedule.HttpPost(id, url, "application/json", lua.json.encode(content));
			return;
			___COROUTINE();
		}).bind(this));
	},
	HandleHttpSucceed : function(id, response) {
		this._id_creator.ReleaseID(id);
		let thread = this._request_map.get(id);
		this._request_map.delete(id);
		let [result, reason] = lua.coroutine.resume(thread, undefined, response);
		if (result !== true) {
			ALittle.Error(reason);
		}
	},
	HandleHttpFailed : function(id, reason) {
		this._id_creator.ReleaseID(id);
		let thread = this._request_map.get(id);
		this._request_map.delete(id);
		let [result, reason] = lua.coroutine.resume(thread, reason, undefined);
		if (result !== true) {
			ALittle.Error(reason);
		}
	},
}, "ALittle.HttpSystem");

window.A_HttpSystem = ALittle.NewObject(ALittle.HttpSystem);
}